### ISSUE TITLE
Attempt to re-implement the results

### ISSUE THREAD

**Issue by Dahuia** [suinleelab/vit-shapley#16]
Hi, thank you for the great work.

I've been trying to check the Table 4 (Pets dataset) results shown in the paper. But couldn't get similar insertion/deletion scores for the baselines (Attention last & SmoothGrad). What I get for both methods are 0.65 for insertion and 0.4 for deletion, but reported in Table 4, insertion is 0.96 and deletion is 0.80.

I couldn't find the detail about training the classifier and calculating the Insertion/Deletion score in your code. It would be great if you can help provide more information about how you train the classifier and evaluate the insertion/deletion score to get the values in Table 4.

Thank you so much for your time.

I am using pre-trained weights from "https://storage.googleapis.com/vit_models/imagenet21k/ViT-B_16.npz" and the way I fine tune the classifer is  by masking the image-embedding (before send to the encoder) then use CE_loss to constrain. 

Pasted below is the way I do insertion/deletion.
-----------

def insertion_deletion_per_patch(model, image, attribution_map, target_label, mode='insertion'):
    """
    Insertion/Deletion score per patch.
    Args:
        model: model taking (1, 3, H, W)
        image: input image tensor (1, 3, H, W)
        attribution_map: tensor of shape (1, num_patches)
        target_label: int, class index
        mode: 'insertion' or 'deletion'
    Returns:
        scores: list of confidence values, one per patch
    """
    image = image.unsqueeze(0)
    image = image.clone()
    _, C, H, W = image.shape
    num_patches = attribution_map.shape[1]
    patch_size = int((H * W / num_patches) ** 0.5)
    saliency = attribution_map.squeeze().detach().cpu().numpy()
    sorted_idx = saliency.argsort()[::-1]  # from most to least important

    if mode == 'insertion':
        current = torch.zeros_like(image)
    elif mode == 'deletion':
        current = image.clone()
    else:
        raise ValueError("mode must be 'insertion' or 'deletion'")

    scores = []

    for idx in sorted_idx:
        row = (idx // (W // patch_size)) * patch_size
        col = (idx % (W // patch_size)) * patch_size

        if mode == 'insertion':
            current[:, :, row:row+patch_size, col:col+patch_size] = image[:, :, row:row+patch_size, col:col+patch_size]
        elif mode == 'deletion':
            current[:, :, row:row+patch_size, col:col+patch_size] = 0

        out = model(current)
        prob = F.softmax(out, dim=1)[0, target_label].item()
        scores.append(prob)

    return scores

**Comment by chanwkimlab** [comment]
Thanks for your interest in our work.

To clarify, the command-line instructions for training the classifiers are available in [scripts/training_classifier.md](https://github.com/suinleelab/vit-shapley/blob/master/scripts/training_classifier.md).

As for the insertion/deletion implementation, you can find the relevant code in [notebooks/2_1_benchmarking.ipynb](https://github.com/suinleelab/vit-shapley/tree/master/notebooks), specifically under the 4_insert_delete section.

One important detail to note is how we handle patch removal in the ViT: “We achieve this by masking attention values at each self-attention layer, or setting them to a large negative value before applying the softmax operation” (Section 4)
Also, please note that we also fine-tune the model to handle such masked inputs, which is important for getting accurate insertion/deletion curves. We found that after fine-tuning, masking attention or intervening at the input level perform similarly (see Figure 2), but using the fine-tuned model is essential either way.

**Comment by Dahuia** [comment]
Great appreciation for your quick response.
 
So what we are talking about here is "it is essential to fine-tune ViT to handle such masked inputs, after which we apply the previous baselines to get the Insertion/Deletion scores?"
 
Two more questions based on that,
  1. Why this fine-tuning is essential? Is this also the same reason you get descent results on **Random** method? Cause in Table 1. **Random** can achieve 0.951 insertion score is actually a lot.
  2. About mask training, I am doing **mask input** level now, is it going to be a big difference between your training - **mask attention**? I am doubting this is also the reason why I can't get accurate insertion/deletion curves even for old baselines.

Thank you so much for your response again. I really appreciate it.

**Comment by chanwkimlab** [comment]
Thanks, those are good questions!
1. Deep learning models are generally not trained to handle masked inputs. So when we apply masking during evaluation without any finetuning, the model output change might not be due to the actual importance of the masked features but also due to getting unexpected input. That's why more sophisticated solutions are needed, such as finetuning. A more comprehensive discussion of this issue can be found in the "Machine learning models are not coalitional games" section of  https://www.nature.com/articles/s42256-023-00657-x  or in Appendix E of https://jmlr.csail.mit.edu/papers/volume22/20-1316/20-1316.pdf. That said, for ViTs specifically, the issue is somewhat less pronounced—they’ve been shown to be more robust to masking than CNNs. Still, as shown in Figure 2 of our paper, fine-tuning is helpful.
2. Yes, the high insertion score for the Random baseline (0.951 in Table 1) may seem surprising, but this is actually expected. If you see Figure 2, the KL divergence between the original and partial-image predictions only starts to increase significantly once most patches are removed. This means that even random insertion can quickly recover the full-image prediction after just a few patches are added. 
3. There wasn't much difference in performance between input-level masking and attention masking, as shown by the overlapping dotted lines in Figure 2. Both approaches would be valid as long as the model is fine-tuned to handle the specific type of masking you’re using.

Hope this clears things up!

**Comment by Dahuia** [comment]
Thank you so much for your help and patience!!!

Just one more small question, when you evaluate for other baselines, 

1. Use ViT to get the importance score/explanation, then use surrogate/fine-tuned model to get the Insertion/Deletion score.
2. Use the surrogate/fine-tuned model to get the explanation and also feed to the same model to get Insertion/Deletion score.

could you please clarify which way you are using? I am fine-tuning ViT myself and tried both ways to get the Insertion/Deletion scores for other baselines or even random, but what I get is not as high as reported in your paper.

Appreciate it so much. You've been such a great help to me.
